modules/demo/client-server/routes/cudf.js

```diff
@@ -0,0 +1,277 @@
+        temp_boolmask = filterByValue(key, value);
+      }
+    }
+    if (temp_boolmask !== undefined) {
+      boolmask = (boolmask == undefined) ? temp_boolmask : boolmask.logical_and(temp_boolmask);
+    }
+  }
+  var res = '';
+  if (boolmask == undefined) {
+    res = df;
+  } else {
+    res = df.filter(boolmask);
+  }
+  var t1 = new Date().getTime();
+  console.log('Query ', query_dict, ', Time Taken:', t1 - t0, 'ms');
+  return res;
+}
+
+/**
+ * Computes cudf.DataFrame.groupby.
+ *
+ * @param df cudf.DataFrame object
+ * @param by str, column name by which the groupby needs to be computed
+ * @param agg str, aggregate function, one of the following:
+ *   https://rapidsai.github.io/node-rapids/classes/cudf_src.groupbysingle.html#methods
+ * @returns resulting cudf.DataFrame
+ */
+function groupby(df, by, agg) {
+  var t0         = new Date().getTime();
+  const grp      = df.groupBy({by: by});
+  const validAgg = (agg in grp && typeof grp[agg] == 'function');
+  if (!validAgg) { return ({}); }
+  var res = grp[agg]();
+  var t1  = new Date().getTime();
+  console.log('Group by ', by, ', agg:', agg, 'Time Taken:', t1 - t0, 'ms');
+  return res;
+}
+
+// socket connection
+module.exports = (io) => {
+  router.get('/', (req, res) => { res.send('ok'); })
+
+  // SOCKET.IO paths
+  io.on('connection', (socket) => {
+    console.log('client connected');
+
+    socket.on('disconnect', () => { console.log('client disconnected'); });
+
+    socket.on('readMortgageData', (callback) => {
+      if (datasetLoaded == 'mortgage') {
+        console.log('dataset already loaded');
+        socket.emit('data-points-update', df.toArrow().length);
+        callback(true);
+      } else {
+        readMortgageData((cb) => {
+          if (cb == true) {
+            socket.emit('data-points-update', df.toArrow().length);
+            callback(true);
+          }
+        });
+      }
+    });
+
+    socket.on('readUberData', (callback) => {
+      if (datasetLoaded == 'uber') {
+        console.log('dataset already loaded');
+        socket.emit('data-points-update', df.toArrow().length);
+        callback(true);
+      } else {
+        readUberData((cb) => {
+          if (cb == true) {
+            socket.emit('data-points-update', df.toArrow().length);
+            callback(true);
+          }
+        });
+      }
+    });
+
+    socket.on('groupby', (by, agg, return_type, query_dict, callback) => {
+      if (df == undefined) { callback({}); }
+      console.log('\n\n');
+      const res_df = parseQuery(query_dict, by);
+      res          = groupby(res_df, by, agg);
+      var t0       = new Date().getTime();
+      res          = res.sortValues({[by]: {ascending: true, null_order: 'AFTER'}});
+      console.log('Sort Dataframe by ', by, ', Time Taken:', new Date().getTime() - t0, 'ms');
+      if (return_type.type == 'geojson') {
+        callback(
+          convertToGeoJSON(res.toArrow(), by, return_type.properties, return_type.geojsonProp))
```

Comment:
We should be able to send the binary Arrow table to the client and instruct deck.gl to use it as-is.